dfc/tests/integration_test.go

```diff
@@ -0,0 +1,301 @@
+
+	// Step 2
+	m.bucket = TestLocalBucketName
+	err = client.CreateLocalBucket(proxyurl, m.bucket)
+	checkFatal(err, t)
+
+	defer func() {
+		err = client.DestroyLocalBucket(proxyurl, m.bucket)
+		checkFatal(err, t)
+	}()
+
+	if usingSG {
+		sgl = dfc.NewSGLIO(filesize)
+		defer sgl.Free()
+	}
+
+	// Step 3
+	m.wg.Add(1)
+	go func() {
+		primaryCrash(t)
+		m.wg.Done()
+	}()
+
+	// PUT phase is timed to ensure it doesn't finish before primaryCrash() begins
+	time.Sleep(5 * time.Second)
+	tlogf("Putting %d files into bucket %s...\n", num, m.bucket)
+	putRandomFiles(0, seed, filesize, num, m.bucket, t, nil, errch, filenameCh, SmokeDir, SmokeStr, "", true, sgl)
+	selectErr(errch, "put", t, false)
+	close(filenameCh)
+
+	for f := range filenameCh {
+		m.repFilenameCh <- repFile{repetitions: m.numGetsEachFile, filename: f}
+	}
+	m.wg.Wait()
+
+	// Step 4
+	m.wg.Add(m.num*m.numGetsEachFile + 2)
+	go func() {
+		// Failback is timed to ensure it doesn't finish before GETs begin
+		time.Sleep(17 * time.Second)
+		primarySetToOriginal(t)
+		m.wg.Done()
+	}()
+
+	go doRepeatedGets(&m)
+	doReregisterTarget(&m)
+
+	m.wg.Wait()
+	tlogf("%d GET errors before target's local bucket map was updated (expected)\n", m.numGetErrsBefore)
+	if m.numGetErrsAfter > maxNumGetErrs {
+		t.Fatalf("Found %d GET errors (should be no more than %d) after target's local bucket map was updated",
+			m.numGetErrsAfter, maxNumGetErrs)
+	} else {
+		tlogf("Found %d GET errors (should be no more than %d) after target's local bucket map was updated",
+			m.numGetErrsAfter, maxNumGetErrs)
+	}
+}
+
+func doRepeatedGets(m *metadata) {
+	for i := 0; i < 10; i++ {
+		m.semaphore <- struct{}{}
+	}
+	tlogf("Getting each of the %d files %d times from bucket %s...\n", m.num, m.numGetsEachFile, m.bucket)
+
+	// GET is timed so a large portion of requests will happen both before and after the target is re-registered
+	time.Sleep(m.delay)
+	for i := 0; i < m.num*m.numGetsEachFile; i++ {
+		go func() {
+			<-m.semaphore
+			defer func() {
+				m.semaphore <- struct{}{}
+				m.wg.Done()
+			}()
+
+			repFile := <-m.repFilenameCh
+			if repFile.repetitions > 0 {
+				repFile.repetitions -= 1
+				m.repFilenameCh <- repFile
+			}
+			_, _, err := client.Get(proxyurl, m.bucket, SmokeStr+"/"+repFile.filename, nil, nil, false, false)
+			if err != nil {
+				r := atomic.LoadUint64(&(m.reregistered))
+				if r == 1 {
+					atomic.AddUint64(&(m.numGetErrsAfter), 1)
+				} else {
+					atomic.AddUint64(&(m.numGetErrsBefore), 1)
+				}
+			}
+			g := atomic.AddUint64(&(m.getsCompleted), 1)
+			if g%5000 == 0 {
+				tlogf(" %d/%d GET requests completed...\n", g, m.num*m.numGetsEachFile)
+			}
+		}()
+	}
+}
+
+func doReregisterTarget(m *metadata) {
+	err := client.RegisterTarget(m.sid, &(m.smap))
+	checkFatal(err, m.t)
+PollLoop:
+	for i := 0; i < 25; i++ {
+		time.Sleep(time.Second)
+		m.smap = getClusterMap(httpclient, m.t)
+		lbNames, err := client.GetLocalBucketNames(m.targetDirectURL)
```

Comment:
Maybe move this line into the next `if`? It does not makes sense to get bucket names from a target if it is not in smap yet.